source/h1_connection.c

```diff
@@ -228,64 +227,77 @@ static int s_stream_send_response(struct aws_http_stream *stream, struct aws_htt
-    }
+    /* Connection window should match the available space in the read-buffer */
+    AWS_ASSERT(connection->thread_data.read_buffer.unprocessed_bytes <= connection->thread_data.read_buffer.capacity);
+    const size_t desired_connection_window =
+        connection->thread_data.read_buffer.capacity - connection->thread_data.read_buffer.unprocessed_bytes;
 
-    /* Otherwise, schedule a task to do it.
-     * If task is already scheduled, just increase size to be updated */
+    AWS_LOGF_TRACE(
+        AWS_LS_HTTP_CONNECTION,
+        "id=%p: Window stats: connection=%zu+%zu stream=%" PRIu64 " buffer=%zu/%zu",
+        (void *)&connection->base,
+        connection->thread_data.window_size,
+        desired_connection_window - connection->thread_data.window_size /*increment_size*/,
+        connection->thread_data.incoming_stream ? connection->thread_data.incoming_stream->thread_data.window_size : 0,
+        connection->thread_data.read_buffer.unprocessed_bytes,
+        connection->thread_data.read_buffer.capacity);
 
-    /* BEGIN CRITICAL SECTION */
-    aws_h1_connection_lock_synced_data(connection);
+    return desired_connection_window;
+}
 
-    /* if this is not volatile, gcc-4x will load window_update_size's address into a register
-     * and then read it as should_schedule_task down below, which will invert its meaning */
-    volatile bool should_schedule_task = (connection->synced_data.window_update_size == 0);
-    connection->synced_data.window_update_size =
-        aws_add_size_saturating(connection->synced_data.window_update_size, increment_size);
+/* Increment connection window, if necessary */
+static int s_update_connection_window(struct aws_h1_connection *connection) {
+    AWS_ASSERT(aws_channel_thread_is_callers_thread(connection->base.channel_slot->channel));
 
-    aws_h1_connection_unlock_synced_data(connection);
-    /* END CRITICAL SECTION */
+    if (connection->thread_data.is_reading_stopped) {
+        return AWS_OP_SUCCESS;
+    }
 
-    if (should_schedule_task) {
-        AWS_LOGF_TRACE(
-            AWS_LS_HTTP_CONNECTION, "id=%p: Scheduling connection cross-thread work task.", (void *)connection_base);
-        aws_channel_schedule_task_now(connection->base.channel_slot->channel, &connection->cross_thread_work_task);
-    } else {
-        AWS_LOGF_TRACE(
-            AWS_LS_HTTP_CONNECTION,
-            "id=%p: Connection cross-thread work task was already scheduled",
-            (void *)connection_base);
+    const size_t desired_size = connection->thread_data.has_switched_protocols
+                                    ? s_calculate_midchannel_desired_connection_window(connection)
+                                    : s_calculate_stream_mode_desired_connection_window(connection);
+
+    const size_t increment_size = aws_sub_size_saturating(desired_size, connection->thread_data.window_size);
+    if (increment_size > 0) {
+        connection->thread_data.window_size += increment_size;
```

Comment:
Should we not record this until AFTER we successfully update the window on the channel?

Reply:
My logic here works by calculating the `desired` window, and issuing a request for the difference between `desired` and `current`.

The `window_size` I'm updating here is NOT the official window according to the channel (that would be `aws_channel_slot.window_size`). That variable isn't updated until the channel runs the window update task. If I based my math off that variable, I could end up asking for too much because there are outstanding updates that haven't landed yet.

I'll add comments explaining this...

Reply:
I had the thought that we might want to modify how `aws_channel_slot.window_size` works, so every channel handler doesn't feel the need to duplicate it for its own internal math.

Like, keep the task that informs handlers about downstream updates, but 

Like, keep using a task to inform a handler's neighbor that its window has been updated, but keep `aws_channel_slot.window_size` as up to date as possible